packages/react/src/analytics/integrations/shared/GoogleConsentMode/GoogleConsentMode.ts

```diff
@@ -0,0 +1,124 @@
+  private dataLayers!: Array<string>; // Stores different data layer names
+  private enabled!: boolean; // enable or disable CMP integration
+  private consentCategoriesMapping?: GoogleConsentModeMappings; // Stores default or customized consent category mappings
+
+  constructor() {
+    this.dataLayers = [];
+    this.enabled = true;
+  }
+
+  /**
+   * Adds an integration for a specific data layer with a given consent.
+   *
+   * @param consent - The consent data.
+   * @param dataLayer - The data layer name.
+   */
+  addIntegration(consent: ConsentData | null, dataLayer: string) {
+    if (this.enabled) {
+      this.dataLayers.push(dataLayer);
+      this.gtag(dataLayer, consent);
+    }
+  }
+
+  /**
+   * Sets custom mappings for consent categories.
+   *
+   * @param consentMapping - Custom mappings to be set.
+   */
+  setConsentCategoriesMapping(consentMapping: GoogleConsentModeMappings) {
+    this.consentCategoriesMapping = consentMapping;
+  }
+
+  /**
+   * Disables Google Consent Mode.
+   */
+  disable() {
+    this.enabled = false;
+  }
+
+  /**
+   * Retrieves consent values for each category based on provided consents.
+   *
+   * @param consents - The consent data.
+   *
+   * @returns Consent values for each category.
+   */
+  private getConsentValues(consents: ConsentData) {
+    // Dealing with null or undefined consent values
+    const consentList = consents || {};
+
+    if (this.consentCategoriesMapping) {
+      return Object.keys(this.consentCategoriesMapping).reduce(
+        (result, consentKey) => ({
+          ...result,
+          [consentKey]: this.consentCategoriesMapping?.[consentKey]?.every(
+            consent => consentList[consent],
+          )
+            ? 'granted'
+            : 'deny',
+        }),
+        {},
+      );
+    }
+
+    // Return default google consent values as deny all
+    return {
+      ad_storage: 'deny',
+      ad_user_data: 'deny',
+      ad_personalization: 'deny',
+      analytics_storage: 'deny',
+    };
+  }
+
+  /**
+   * Sets consents for specified data layers.
+   *
+   * @param consents - The consent data to be set.
+   * @param dataLayer - The specific data layer where consents should be set.
+   */
+  setConsents(consents: ConsentData, dataLayer?: string) {
+    if (this.enabled) {
+      const dataLayers = dataLayer ? [dataLayer] : this.dataLayers;
+
+      dataLayers.forEach(layer => this.gtag(layer, consents, 'update'));
+    }
+  }
+
+  /**
+   * Handles consent by updating the data layer with consent information.
+   *
+   * @param dataLayer - The name of the data layer.
+   * @param consent - The consent data.
+   * @param command - Optional. The command for consent handling.
+   */
+  private gtag(
+    dataLayer: string,
+    consent: ConsentData | null,
+    command = 'default',
+  ) {
+    if (this.enabled && typeof window !== 'undefined') {
+      // @ts-ignore
+      window[dataLayer] = window[dataLayer] || [];
+
+      function internalGtag() {
```

Comment:
Why is it necessary to have this function? I think this will complicate the code a little bit and the performance will be affected since this function will always be recreated when the gtag method is called.